README.txt
=================
silva.app.sitemap
=================

Introduction
============

``silva.app.sitemap`` is an extension let you generate xml sitemaps
0.9 (see http://www.sitemaps.org/protocol.html).

A view ``sitemap.xml`` will be available on every folder of your site
in order to access the sitemap.

A local service will permit you to configure which contetnt you want to
see in the sitemap.


Code repository
===============

Source code can be found at:
https://hg.infrae.com/silva.app.sitemap


docs/HISTORY.txt
Changes
=======

1.0.1 (unreleased)
------------------

* ...

1.0 (2012/12/11)
----------------

* Initial release


src/silva/app/sitemap/tests/test_service.py

import unittest

from zope import component

from ..testing import FunctionalLayer
from ..interfaces import ISitemapService

from silva.core.interfaces import IAccessSecurity


class ExcludedContentTestCase(unittest.TestCase):

    layer = FunctionalLayer

    def setUp(self):
        self.root = self.layer.get_application()
        self.service = component.getUtility(ISitemapService)
        factory = self.root.manage_addProduct['Silva']
        factory.manage_addFolder('public', 'Public')
        factory.manage_addFolder('private', 'Private')

    def test_excluded_folder(self):
        private_path = self.root.private.getPhysicalPath()
        public_path = self.root.public.getPhysicalPath()

        self.assertFalse(self.service.is_path_excluded(public_path))
        self.assertFalse(self.service.is_path_excluded(private_path))
        self.assertFalse(self.service.is_path_excluded(
            private_path + ('subcomponent',)))
        self.assertFalse(self.service.is_path_excluded(
            self.root.getPhysicalPath()))

        self.service.exclude_content(self.root.private)
        self.assertFalse(self.service.is_path_excluded(
            self.root.getPhysicalPath()))
        self.assertFalse(self.service.is_path_excluded(public_path))
        self.assertTrue(self.service.is_path_excluded(
            private_path))
        self.assertTrue(self.service.is_path_excluded(
            private_path + ('subcomponent',)))

        self.service.unexclude_content(self.root.private)
        self.assertFalse(self.service.is_path_excluded(
            private_path))
        self.assertFalse(self.service.is_path_excluded(
            private_path + ('subcomponent',)))
        self.assertFalse(self.service.is_path_excluded(
            self.root.getPhysicalPath()))

    def test_container_excluded_when_rescriction_set(self):
        private_path = self.root.private.getPhysicalPath()
        self.assertFalse(self.service.is_path_excluded(private_path))
        IAccessSecurity(self.root.private).set_minimum_role('Authenticated')
        self.assertTrue(self.service.is_path_excluded(private_path))
        self.assertTrue(self.service.is_path_excluded(
            private_path + ('subcomponent',)))

        IAccessSecurity(self.root.private).set_minimum_role('Anonymous')
        self.assertFalse(self.service.is_path_excluded(private_path))


def test_suite():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(ExcludedContentTestCase))
    return suite


src/silva/app/sitemap/tests/test_sitemap_views.py
# -*- coding: utf-8 -*-

import unittest

from zope.component import queryUtility
from zope.event import notify
from zope.lifecycleevent import ObjectModifiedEvent

from Products.Silva.testing import FunctionalLayer as SilvaFunctionalLayer
from silva.core.interfaces import IPublicationWorkflow
from ..interfaces import ISitemapService
from ..testing import FunctionalLayer


class SitemapNotInstalled(unittest.TestCase):
    layer = SilvaFunctionalLayer

    def setUp(self):
        self.root = self.layer.get_application()

    def test_view_not_found(self):
        with self.layer.get_browser() as browser:
            self.assertEquals(browser.open('/root/sitemap.xml'), 404)


class SitemapViewTestCase(unittest.TestCase):
    layer = FunctionalLayer

    def setUp(self):
        self.root = self.layer.get_application()
        self.layer.login('manager')
        factory = self.root.manage_addProduct['Silva']
        factory.manage_addFolder('folder', 'Folder')
        factory = self.root.folder.manage_addProduct['Silva']
        factory.manage_addMockupVersionedContent('index', 'Index')
        notify(ObjectModifiedEvent(self.root.folder))
        IPublicationWorkflow(self.root.folder.index).publish()
        self.assertTrue(self.root.folder.is_published())

    def test_simple_view(self):
        service = queryUtility(ISitemapService)
        service.set_trim_index(False)
        service.set_allowed_meta_types([
            'Mockup VersionedContent'])

        with self.layer.get_browser() as browser:
            self.assertEquals(browser.open('/root/sitemap.xml'), 200)

    def test_view_on_folder(self):
        service = queryUtility(ISitemapService)
        service.set_trim_index(False)
        service.set_allowed_meta_types([
            'Mockup VersionedContent'])

        with self.layer.get_browser() as browser:
            self.assertEquals(browser.open('/root/folder/sitemap.xml'), 200)


def test_suite():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(SitemapNotInstalled))
    suite.addTest(unittest.makeSuite(SitemapViewTestCase))
    return suite


src/silva/app/sitemap/tests/test_buffered_streaming.py

import unittest
from random import randint
from ..views import BufferedGeneratorStream


class BufferedStreamingTest(unittest.TestCase):

    def test_data_smaller_than_buf_size(self):
        gen = iter(['abcd'])

        streamer = BufferedGeneratorStream(gen, size=100)
        self.assertEquals(['abcd'], list(streamer))

    def test_data_bigger_tan_buf_size(self):
        """ each item is bigger that buf size so each item is passed through.
        """
        gen = iter(['a'*20 for i in range(10)])
        streamer = BufferedGeneratorStream(gen, size=10)
        self.assertEquals(['a'*20 for i in range(10)], list(streamer))

    def test_big_stream(self):
        data = ['ab' * randint(100, 150) for i in range(1000)]
        gen = iter(data)
        streamer = BufferedGeneratorStream(gen, size=4*1024)
        self.assertEquals("".join(streamer),
                          "".join(data))

    def test_data_is_multiple_of_buf_size(self):
        data = ['ab' * 1024 for i in range(100)]
        gen = iter(data)
        streamer = BufferedGeneratorStream(gen, size=2048)
        chunks = list(streamer)
        self.assertEquals(100, len(chunks))
        self.assertEquals(data, chunks)


def test_suite():
    suite = unittest.TestSuite()
    suite.addTest(unittest.makeSuite(BufferedStreamingTest))
    return suite


src/silva/app/sitemap/service.py
# Copyright (c) 2012 Infrae. All rights reserved.
# See also LICENSE.txt


from five import grok

from zope import component
from zope import schema
from zope.interface import Interface
from zope.lifecycleevent import IObjectAddedEvent
from zope.schema.interfaces import IContextSourceBinder
from zope.schema.vocabulary import SimpleVocabulary, SimpleTerm

from silva.core import conf as silvaconf
from silva.core.interfaces import IAccessSecurity, IContainer
from silva.core.interfaces import ISecurityRestrictionModifiedEvent
from silva.core.interfaces.adapters import IAddableContents
from silva.core.interfaces.content import IPublishable
from silva.core.services.base import SilvaService
from silva.core.services.utils import walk_silva_tree

from .interfaces import ISitemapService
from zeam.form import silva as silvaforms
from zeam.form.ztk.actions import EditAction

from AccessControl import ClassSecurityInfo
from Acquisition import aq_parent
from App.class_init import InitializeClass

from .utils import TupleMap


class SitemapService(SilvaService):
    grok.name('service_sitemap')
    grok.implements(ISitemapService)
    meta_type = "Silva Sitemap Service"
    silvaconf.icon('static/sitemap.png')

    security = ClassSecurityInfo()

    manage_options = (
        {'label': 'Edit', 'action': 'manage_settings'},
        ) + SilvaService.manage_options

    _allowed_meta_types = None
    _excluded_paths = None
    _excluded_contents = None
    _paths = None
    _trim_index = True

    def __init__(self, id):
        super(SitemapService, self).__init__(id)
        self._allowed_meta_types = set()
        self._excluded_paths = set()
        self._paths = set()
        self._excluded_contents = TupleMap()

    security.declareProtected('View Management Screens', 'exclude_content')
    def exclude_content(self, content):
        path = content.getPhysicalPath()
        try:
            self._excluded_contents.add(path, True)
            self._p_changed = 1
        except KeyError:
            pass

    security.declareProtected('View Management Screens', 'unexclude_content')
    def unexclude_content(self, content):
        path = content.getPhysicalPath()
        try:
            self._excluded_contents.remove(path)
            self._p_changed = 1
        except KeyError:
            pass

    security.declareProtected('View Management Screens', 'build')
    def build(self):
        self._excluded_contents = TupleMap()
        for container in walk_silva_tree(self.get_root(), requires=IContainer):
            self.index_content_access(container)

    security.declareProtected('View Management Screens', 'index_content_access')
    def index_content_access(self, content):
        access = IAccessSecurity(content)
        if access.get_minimum_role() is None:
            self.unexclude_content(content)
        elif not access.is_acquired():
            self.exclude_content(content)

    def is_path_excluded(self, path):
        path, index = self._excluded_contents.get(path, fallback=True)
        return path is not None

    def get_trim_index(self):
        return self._trim_index

    security.declareProtected(
        'View Management Screens', 'set_trim_index')
    def set_trim_index(self, value):
        self._trim_index = value

    def get_allowed_meta_types(self):
        return self._allowed_meta_types

    security.declareProtected(
        'View Management Screens', 'set_allowed_meta_types')
    def set_allowed_meta_types(self, meta_types):
        self._allowed_meta_types = set(meta_types)
        # get addables IAllowedAddables

    def get_excluded_paths(self):
        return self._excluded_paths

    security.declareProtected(
        'View Management Screens', 'set_excluded_paths')
    def set_excluded_paths(self, excluded_paths):
        self._excluded_paths = set(excluded_paths)

    def get_paths(self):
        return self._paths

    security.declareProtected(
        'View Management Screens', 'set_paths')
    def set_paths(self, paths):
        self._paths = set(paths)


InitializeClass(SitemapService)


@grok.subscribe(IContainer, ISecurityRestrictionModifiedEvent)
def register_excluded_container(container, event):
    service = component.queryUtility(ISitemapService)
    if service is not None:
        service.index_content_access(container)


@grok.subscribe(ISitemapService, IObjectAddedEvent)
def set_defaults(service, event):
    service.set_paths(['/'.join(aq_parent(service).getPhysicalPath())])
    addables = IAddableContents(
        aq_parent(service)).get_all_addables(IPublishable)
    service.set_allowed_meta_types(addables)


@grok.provider(IContextSourceBinder)
def addables_content_types(context):
    addables = IAddableContents(aq_parent(context)).get_all_addables()
    entries = []
    for addable in addables:
        entries.append(SimpleTerm(value=addable,
                                  token=addable,
                                  title=addable))
    return SimpleVocabulary(entries)


class ISitemapFields(Interface):
    _allowed_meta_types = schema.Set(
        title=u'Allowed Meta types',
        value_type=schema.Choice(source=addables_content_types),
        required=True)
    _paths = schema.Set(
        title=u'Allowed paths',
        value_type=schema.TextLine(),
        required=True)
    _trim_index = schema.Bool(
        title=u'Remove index from path',
        description=u'When a content is index of a folder remove the '
        u'/index part of the url.')


class SitemapSettings(silvaforms.ZMIComposedForm):
    grok.name('manage_settings')
    grok.context(ISitemapService)
    label = 'Manage sitemap service'


class SitemapConfigurationForm(silvaforms.ZMISubForm):
    grok.context(ISitemapService)
    grok.view(SitemapSettings)
    label = 'Settings'
    ignoreContent = False
    fields = silvaforms.Fields(ISitemapFields)
    actions = silvaforms.Actions(EditAction('Save changes'))


class SitemapRebuildForm(silvaforms.ZMISubForm):
    grok.context(ISitemapService)
    grok.view(SitemapSettings)

    label = 'Rebuild index'
    description = ('Build the index of excluded contents'
                   ' from access restrictions.')

    @silvaforms.action('Rebuild')
    def rebuild(self):
        self.status = 'Index built.'
        self.context.build()


src/silva/app/sitemap/__init__.py
# Copyright (c) 2012 Infrae. All rights reserved.
# See also LICENSE.txt

from zope.interface import Interface

from silva.core import conf as silvaconf
from silva.core.conf.installer import DefaultInstaller

silvaconf.extension_name("silva.app.sitemap")
silvaconf.extension_title("Silva Sitemaps")
silvaconf.extension_depends(['Silva'])


class SilvaSitemapInstaller(DefaultInstaller):
    """ Silva sitemap installer
    """

    service_name = 'service_sitemap'

    def install_custom(self, root):
        if self.service_name not in root.objectIds():
            factory = root.manage_addProduct['silva.app.sitemap']
            factory.manage_addSitemapService(self.service_name)


class IExtension(Interface):
    """silva.app.sitemap extension
    """


install = SilvaSitemapInstaller("silva.app.sitemap", IExtension)


src/silva/app/sitemap/views.py
# Copyright (c) 2012 Infrae. All rights reserved.
# See also LICENSE.txt

from xml.sax.saxutils import escape
from itertools import imap, ifilter

from five import grok
from zope.component import getUtility, queryUtility
from zope.traversing.browser.absoluteurl import absoluteURL
from zope.publisher.interfaces.http import IResult

from zExceptions import NotFound

from silva.core.interfaces import IContainer
from silva.core.services.interfaces import ICatalogService

from .interfaces import ISitemapService


class BufferedGeneratorStream(object):
    grok.implements(IResult)

    def __init__(self, generator, size=200 * 1024, encoding='utf-8'):
        self.generator = generator
        self.encoding = encoding
        self._buffer = bytearray(size)
        self._cursor = 0
        self._size = size
        self._done = False

    def __iter__(self):
        return self

    def next(self):
        if self._done:
            raise StopIteration

        try:
            remaining = ''
            while self._cursor < self._size:
                item = self.generator.next().encode(self.encoding)
                remaining_len = self._size - self._cursor
                if len(item) > remaining_len:
                    remaining = item
                    break
                next_pos = self._cursor + len(item)
                self._buffer[self._cursor:next_pos] = item
                self._cursor = next_pos

            result = str(self._buffer[:self._cursor]) + remaining
            self._cursor = 0
            return result
        except StopIteration:
            self._done = True
            if self._cursor:
                return str(self._buffer[:self._cursor])
            raise StopIteration


def _normalized_path(p):
    return p.endswith('/') and p or p + '/'

def _utf8_encode(s):
    return s.encode('utf-8')


class SitemapView(grok.View):
    grok.context(IContainer)
    grok.name('sitemap.xml')

    def update(self):
        self.sitemap = queryUtility(ISitemapService)
        if self.sitemap is None:
            raise NotFound()
        catalog = getUtility(ICatalogService)
        query = dict(
            path=map(_utf8_encode, self.sitemap.get_paths()),
            publication_status="public",
            meta_type=map(_utf8_encode, self.sitemap.get_allowed_meta_types())
        )
        brains = catalog.search(query)
        self.trim_index = self.sitemap.get_trim_index()
        self.seen = set()
        self.excluded_paths = map(_normalized_path,
            self.sitemap.get_excluded_paths())

        pipe = iter(brains)
        pipe = ifilter(self.has_date, pipe)
        pipe = ifilter(self.exclude, pipe)
        pipe = imap(self.process, pipe)
        pipe = ifilter(self.dedup, pipe)
        self.pipe = pipe

    def exclude(self, brain):
        return not self.sitemap.is_path_excluded(brain.getPath().split('/'))

    def has_date(self, brain):
        return brain['silva-extramodificationtime'] is not None

    def dedup(self, item):
        url = item[0]
        if url in self.seen:
            return False
        self.seen.add(url)
        return True

    def process(self, brain):
        url = str(absoluteURL(brain, self.request))
        if self.trim_index and url.endswith('/index'):
            url = url[0:-6]
        return (url, brain['silva-extramodificationtime'].HTML4())

    def render(self):
        self.request.response.setHeader(
            'Content-Type', 'text/xml;charset=utf-8')
        return BufferedGeneratorStream(self.generate_xml())

    def generate_xml(self):
        yield ('<?xml version="1.0" encoding="UTF-8"?>\n'
               '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">')
        for count, r in enumerate(self.pipe):
            if count > 50000:
                break
            yield """
<url>
   <loc>%s</loc>
   <lastmod>%s</lastmod>
</url>""" % (escape(r[0]), r[1])
        yield '\n</urlset>\n'
